### PR TITLE
Feature/application help centralising config

### PR DESCRIPTION
**Pull request by mod**

### CODE REVIEW

File: lib/peatio/config.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+module Peatio
+  class Config < OpenStruct
+
+    def method_missing(sym, *args, &blk)
```
Comment (seletskiy) [inline lib/peatio/config.rb]: Why so magic? Prefer to have config structure with strictly defined fields along with required and optional values specified as Hash.

Right now when developer look at Config he has no idea about which configuration values are actually used by gem.

**Comment by mod** [comment]
Your both comments are valid and it's something we should keep in mind:
- Avoid application centralising logic but use it more as a middleware Like pattern like Rails and Railtie in my case services are sub-Applications

- There is a chance we can remove the Singleton state and do 1 Root instance with it's Service child decorator.

- We can leave a little bit of magic for the sake of simplicity, the configuration definition is in config/application.rb not lib folder.

I will not be able to have time to update this patch from now as I am not in holiday, I suggest we merge it and you can refactor on top of it in order to not make this patch any bigger.

This is also the Ruby way.

File: lib/peatio/config.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+      name = sym.to_s.gsub(/[=!?]$/, '').to_sym
+      value = super
+      if value.nil?
+        value = Config.new
+        self[name] = value
+      end
+      return value
```
Comment (rxx) [inline lib/peatio/config.rb]: You can write just `value`

File: lib/peatio/application.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+      File.expand_path('../..', File.dirname(__FILE__))
+    end
+
+    def self.config_path(env)
+      filename = '%s.rb' % [env]
+      config_path = File.join(self.path, 'config', filename)
+      unless File.exist?(config_path)
+        raise Error.new(code: 100, text: "Invalid config path #{config_path}")
+      end
+      return config_path
```
Comment (rxx) [inline lib/peatio/application.rb]: I suggest to use earlier `return`
```
return config_path if File.exist?(config_path)
raise Error.new(code: 100, text: "Invalid config path #{config_path}")
```

**Comment by rxx** [comment]
I like the idea and I don't think there is a lot of magic. That's all much clear to understand

File: lib/peatio/application.rb (hunk 3 above)
Comment (seletskiy) [reply lib/peatio/application.rb]: @rxx Isn't it more idiomatic to return early for errors (I think it's called guard clauses).

**Comment by seletskiy** [comment]
@rxx: Well, in my experience explicit always better than implicit in terms of code maintainability even if in result you will get slightly more code.

As I already stated, looking on the config class it's now impossible to know which configuration parameters application use/expect which raises overall cognitive complexity for developer who is not familiar with code. We write programs to be easy to understand by people, right?

**Comment by kovetskiy** [comment]
@mod No, I don't want to merge this patch as is. There is no need of environments right now, it's still not integrated to peatio, I want to keep the code clean and simple as possible. We don't need God Object here, we can just apply our patches on top of yours.

**Pull request closed without merging**